nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/CoreInstance.java

```diff
@@ -329,4 +341,66 @@ public void run() {
         }.runUnrestricted();
     }
 
+    /**
+     * Performs a {@link Function} atomically, in a cluster-wide critical section.
+     *
+     * @param session an existing session
+     * @param obj the object on which the function should be applied
+     * @param keyFunction the function used to generate the key for locking/unlocking the object
+     * @param function the function to apply while the object is locked
+     * @return the result of the function
+     * @since 9.3
+     */
+    public static <T, R> R doAtomically(CoreSession session, T obj, BiFunction<CoreSession, T, String> keyFunction,
+            BiFunction<CoreSession, T, R> function) throws ConcurrentUpdateException, InterruptedException {
```

Comment:
The signature can be simplified:
- instead of `keyFunction` just pass the key directly, there's no benefit in computing it here, the caller can do it.
- instead of `session` + `obj` + `function` just pass a `Supplier<R>`. The caller can capture all the logic it needs in a lambda
- also provide an alternative signature for `doAtomically` that takes a `Callable` instead of a `Supplier<R>` and returns void (see `TransactionHelper.runInTransaction` for a similar pattern).